File: examples2d/polyline2.rs
use rapier_testbed2d::Testbed;
use rapier2d::na::ComplexField;
use rapier2d::prelude::*;

pub fn init_world(testbed: &mut Testbed) {
    /*
     * World
     */
    let mut world = PhysicsWorld::new();

    /*
     * Ground
     */
    let ground_size = 50.0;
    let nsubdivs = 2000;
    let step_size = ground_size / (nsubdivs as f32);
    let mut points = Vec::new();

    points.push(Vector::new(-ground_size / 2.0, 40.0));
    for i in 1..nsubdivs - 1 {
        let x = -ground_size / 2.0 + i as f32 * step_size;
        let y = ComplexField::cos(i as f32 * step_size) * 2.0;
        points.push(Vector::new(x, y));
    }
    points.push(Vector::new(ground_size / 2.0, 40.0));

    let rigid_body = RigidBodyBuilder::fixed();
    let collider = ColliderBuilder::polyline(points, None);
    let _ = world.insert(rigid_body, collider);

    /*
     * Create the cubes
     */
    let num = 20;
    let rad = 0.5;

    let shift = rad * 2.0;
    let centerx = shift * (num / 2) as f32;
    let centery = shift / 2.0;

    for i in 0..num {
        for j in 0usize..num {
            let x = i as f32 * shift - centerx;
            let y = j as f32 * shift + centery + 3.0;

            // Build the rigid body.
            let rigid_body = RigidBodyBuilder::dynamic().translation(Vector::new(x, y));

            if j % 2 == 0 {
                let collider = ColliderBuilder::cuboid(rad, rad);
                let _ = world.insert(rigid_body, collider);
            } else {
                let collider = ColliderBuilder::ball(rad);
                let _ = world.insert(rigid_body, collider);
            }
        }
    }

    /*
     * Set up the testbed.
     */
    testbed.set_physics_world(world);
    testbed.look_at(Vec2::ZERO, 10.0);
}

File: python/rapier-testbed/rapier_testbed/examples2/polyline2.py
"""Port of examples2d/polyline2.rs."""
from __future__ import annotations

import math

import numpy as np

import rapier2d as rp
from .._registry import register

CATEGORY = "Collisions"
NAME = "Polyline"


def init_world(testbed) -> None:
    bodies = rp.RigidBodySet()
    colliders = rp.ColliderSet()
    impulse_joints = rp.ImpulseJointSet()
    multibody_joints = rp.MultibodyJointSet()

    ground_size = 50.0
    nsubdivs = 2000
    step_size = ground_size / nsubdivs

    pts = [[-ground_size / 2.0, 40.0]]
    for i in range(1, nsubdivs - 1):
        x = -ground_size / 2.0 + i * step_size
        y = math.cos(i * step_size) * 2.0
        pts.append([x, y])
    pts.append([ground_size / 2.0, 40.0])

    vertices = np.array(pts, dtype=np.float32)
    handle = bodies.insert(rp.RigidBody.fixed())
    colliders.insert_with_parent(rp.Collider.polyline(vertices), handle, bodies)

    num = 20
    rad = 0.5
    shift = rad * 2.0
    centerx = shift * (num // 2)
    centery = shift / 2.0

    for i in range(num):
        for j in range(num):
            x = i * shift - centerx
            y = j * shift + centery + 3.0
            handle = bodies.insert(rp.RigidBody.dynamic().translation((x, y)))
            if j % 2 == 0:
                colliders.insert_with_parent(
                    rp.Collider.cuboid(rad, rad), handle, bodies
                )
            else:
                colliders.insert_with_parent(rp.Collider.ball(rad), handle, bodies)

    testbed.set_world(bodies, colliders, impulse_joints, multibody_joints)
    testbed.set_camera_2d(center=(0.0, 0.0), zoom=10.0)


register(CATEGORY, NAME, init_world, dim=2)

if __name__ == "__main__":
    from .._picker import run

    run(initial=f"{CATEGORY} / {NAME}")
